### ISSUE TITLE
MySensors SmartSleep (again)

### ISSUE THREAD

**Issue by MarcusKlerks** [domoticz/domoticz#2533]
With issue #2487 the implementation of MySensors SmartSleep has been corrected and reported as ok. I have to get back to this since the working is not reliable. After the update everything went fine until suddenly it stopped working. The only thing that happened was an update of Domoticz on my system. Clearly it had nothing to do with the update itself (nothing touched MySensors) but, confirmed with testing, the restart of Domoticz causes Smartsleep to stop working.
I tried to perform some debugging:
- No difference in database data (MySensors tables) with working and non-working Smartsleep
- No difference in Domoticz logging (debug on) except non-working Smartsleep gives error messages (Error: MySensors: Repeating previous command (2/2) and 
Error: MySensors: Command not received by Node !! (node_id: 101, child_id: 1))
- Other children in the node, not reacting on SmartSleep continue to work (so the node is alive).
- Debugging at the node shows that PRE-NOTIFICATION and POST-NOTIFICATION messages are present.

Do you have any clue to what can cause this behaviour? Is there anything else I can look at?

Thanks, Mark

**Comment by MarcusKlerks** [comment]
Additional, to correct the situation it is necessary to delete the node and device(s) in Domoticz and erase the e-prom (and thus sketch) at the node. Then reload the sketch and connect to Domoticz. The SmartSleep functionality will keep working till the next restart of Domoticz :(  Something wrong during initialization of MySensors hardware in Domoticz?

Mark

**Comment by jacfont** [comment]
Hello,
I'm running in the same issue ( DomoticzV4.10717 & mySensor 2.3.2)
The solution to delete the node & erase the eeprom did the trick.
Any further update / idea on what's the problem ?
Jacques

**Comment by gizmocuz** [comment]
Are you able to debug Domoticz in combination with your sensor ?
Like compiling under windows and setting break points in visual studio ?

**Comment by jacfont** [comment]
Hello,
thanks for your fast reaction.
My setup is running on RPI and I'm not really a developer. (my C++ experience is limited to arduino)
I could setup a windows machine to build domoticz. I'll try to make time to do this during the week-end.

**Comment by jacfont** [comment]
Hello, I'm available to try to debug Domoticz/SmartSleep. Do you some guideline to describe where to begin ?

**Comment by gizmocuz** [comment]
@jacfontaine , that would be great!
Are you on Windows, then i would suggest to checkout the code, install visual studio 2019, and follow the steps in install.md to get you up and compiling.
Then set breakpoints in the MySensor class

**Comment by jacfont** [comment]
@gizmocuz ,
vs2019 is up and compiled domoticz. I build the mysensors setup again to reproduce the problem.
Very basic : domoticz->mysensors_gw(serial)->nrf24->sensors
To test, I send a update to a text variable trough api.

Without debugging code yet, I can detect the following:
- with a fresh setup work like a charm. 
- after a restart of domoticz, it stop to work. (till a workaround)

i see that the problem is the delay to answer the sleep notification. (i had to update the wait_sleep_duration from 0.5s to 5sec ...)
- fresh : 150msec (to receive the updated info)
-after : 2500msec 

Does this give you already some idea on the issue ?
I can search further. Any idea where to look ?


Some extra info:
the log of my sensor:
- fresh
4771 TSF:TRI:TSB
4780 TSF:MSG:SEND,66-66-0-0,s=255,c=3,t=33,pt=5,l=4,sg=0,ft=0,st=OK:120000
4787 MCO:SLP:MS=120000,SMS=1,I1=255,M1=255,I2=255,M2=255
4794 TSF:MSG:SEND,66-66-0-0,s=255,c=3,t=32,pt=5,l=4,sg=0,ft=0,st=OK:500
4812 TSF:MSG:READ,0-0-66,s=12,c=1,t=47,pt=0,l=8,sg=0:info--04

- after restart domoticz

12338 TSF:MSG:SEND,66-66-0-0,s=255,c=3,t=33,pt=5,l=4,sg=0,ft=0,st=OK:90000
12346 TSF:MSG:SEND,66-66-0-0,s=255,c=3,t=1,pt=0,l=0,sg=0,ft=0,st=OK:
12353 MCO:SLP:MS=90000,SMS=1,I1=255,M1=255,I2=255,M2=255
12362 TSF:MSG:SEND,66-66-0-0,s=255,c=3,t=32,pt=5,l=4,sg=0,ft=0,st=OK:5000
12369 TSF:MSG:READ,0-0-66,s=255,c=3,t=1,pt=0,l=10,sg=0:1587134106
14858 TSF:MSG:READ,0-0-66,s=12,c=1,t=47,pt=0,l=8,sg=0:info--13
14864 TSF:MSG:ECHO REQ
14868 TSF:MSG:SEND,66-66-0-0,s=12,c=1,t=47,pt=0,l=8,sg=0,ft=0,st=OK:info--13
14880 TSF:MSG:READ,0-0-66,s=12,c=1,t=47,pt=0,l=8,sg=0:info--13
14892 TSF:MSG:ECHO REQ
14896 TSF:MSG:SEND,66-66-0-0,s=12,c=1,t=47,pt=0,l=8,sg=0,ft=0,st=OK:info--13

- domoticz log:

2020-04-17 14:35:06.376  MySensors: NodeID: 66, ChildID: 255, MessageType: 3, Ack: 0, SubType: 33, Payload: 90000
2020-04-17 14:35:06.384  MySensors: NodeID: 66, ChildID: 255, MessageType: 3, Ack: 0, SubType: 1, Payload:
2020-04-17 14:35:06.385  Status: MySensors: going to send: 66;255;3;0;1;1587134106

2020-04-17 14:35:06.462  MySensors: NodeID: 66, ChildID: 255, MessageType: 3, Ack: 0, SubType: 32, Payload: 5000
2020-04-17 14:35:06.462  Status: MySensors: going to send: 66;12;1;1;47;info--13

2020-04-17 14:35:07.674  Error: MySensors: Repeating previous command (2/2)
2020-04-17 14:35:07.674  Status: MySensors: going to send: 66;12;1;1;47;info--13

2020-04-17 14:35:08.886  Error: MySensors: Command not received by Node !! (node_id: 66, child_id: 12)
2020-04-17 14:35:08.963  MySensors: NodeID: 66, ChildID: 12, MessageType: 1, Ack: 1, SubType: 47, Payload: info--13
2020-04-17 14:35:08.965  MySensors: NodeID: 66, ChildID: 12, MessageType: 1, Ack: 1, SubType: 47, Payload: info--13
2020-04-17 14:35:09.009  (mySensors_GW) General/Text (test4 - Buzzer)



A side point, I receive all message 2x. Not a big deal. (not with the fresh setup)

**Comment by markcame** [comment]
In addition i see that if you perform this sequence you can recover the fresh response

1 . On the hardware setup of mysensor select the node that use smartsleep then delete it
2.  Disable and re-enable Mysensor Hardware
3.  Send a re-presentation with a reboot of the node

If you do not perform one of the steps you are not able to recover, this until next reboot or restart of domoticz process

Hope this can be useful to debug the problem

**Comment by jacfont** [comment]
Hello @gizmocuz ,

I did a bit of debug in the code. Here are my findings:
- the notification 'I_PRE_SLEEP_NOTIFICATION' is well recognized and handled. (in MySensorsBase::ParseLine)
- the sleep_queue is managed and the message is push to the sendQueue (trough SendNodeSetCommandImpl /  SendCommandInt  / m_sendQueue.push(sstr.str());
)

The problem could be the async thread of msg processing.
I see a scheduler of 2 sec. (correct ?)

We should find a way to fire the scheduler when a msg is push in the sendQueue.

But why is it working better for a fresh node ?
A side question : It seems that after a server restart the smartSleep state is forgotten till next msg is received. Should this be a stored properties of the Node ?

**Comment by gizmocuz** [comment]
@jacfontaine , thank you for looking into this.
I don't use the smartsleep function and am not able to assist in this without creating such a node
You could extend the MySensorChilds table to add the smartsleep state here, but should this be done on the application side ?
From what I remember (a long time ago) a device that is using 'smartsleep' is able to send something to the application and is awake for some seconds to receive a command ?

There is no 2 second scheduler here ... all is send here when this state is received: 
I_PRE_SLEEP_NOTIFICATION

**Comment by markcame** [comment]
@gizmocuz , i've made a bit of trial and the strange 
 With a fresh node till domoticz or hardware restart --> I_PRE_SLEEP_NOTIFICATION, and I_POST_SLEEP_NOTIFICATION are present in log but actually there are no active messages sended from domoticz (for example for a change of state of a switch) but despite this all funcion as expceted because it seems there are a  sync between state of domoticz and node (but i didn't see any message logged).
After the restart of domoticz or hardware the messages are correctly handled by sendQueue 
but there are a terrible delay in sending messages
the behaviour occurs both on my raspberry and also on a fresh installation of domoticz and is always about 2 seconds and sometimes 3 seconds.

in the attached log a sensor with PIR and a led controlled by a switch on domoticz
(if the pir is triggered and the switch on domoticz is on when the node awake the a led light will be turned  on)

as you can see in all the first part of log there are no messages about the switch altought this will be triggered in domoticz and the node turn on and off the light accordingly and with a very fast response, less than a second.

in the last part of the log from line 243, the PIR goes off and the node entering in smartsleep, then the switch was turned on in domoticz and after the node awake form smartsleep triggered by pir, prior to go in sleep send I_PRE_SLEEP_NOTIFICATION but only 2 seconds later domoticz respond twice with a V_STATUS 1, this is a bit annoyng because needs to keep the node awake for many seconds at every trigger


[logmysensors.txt](https://github.com/domoticz/domoticz/files/4500436/logmysensors.txt)

**Comment by markcame** [comment]
Ok Probably i find the problem

in MySensorsBase.cpp in function SendNodeSetCommandImpl (line 1274)
when first entering the funcion 

SendCommandInt(NodeID, ChildID, messageType, bUseAck, SubType, Payload);

was never called because of repat = 0 then we wait for an ack time (1200 ms)
then error was reported on log of domoticz and at the second time with repeat =1  the function was executed but we have to wait for a secon ack time 1200 becouse of the firs fail and then in total we ahave about 2400 time consumed , if you reduce the ack time to 150 ms we obtain a delay of 300 an so on.

probably adding resolve the main issue 
remain to investigate because wait for the ack timeout

`if (repeat == 0)

			_log.Log(LOG_STATUS, "MySensors: Sending command to awake (node_id: %d, child_id: %d)", NodeID, ChildID);

		SendCommandInt(NodeID, ChildID, messageType, bUseAck, SubType, Payload);`

**Comment by markcame** [comment]
With a deeper investigation sems a problem with the ACK, if you disable it the node will execute the command afther the preesleep notification in few milliseconds.
I've verified that the echo messages are sent back from node to domoticz but i seems is not processed or not well processed.
@gizmocuz what is the funcion that return the ack ?
i've also verified  that the funcion 
SendCommandInt(NodeID, ChildID, messageType, bUseAck, SubType, Payload); is correctly invoked frist time and then repeated 2 times more but what happens is that seems not receive the ack and so waits until the timeout expire (1200 ms default by two times).

**Comment by kofec** [comment]
Regarding ACK, I see the same issue. But I wasn't sure if Mysensors node is not sending this ACK or Domoticz incorrectly process it or don't see it at all. So in my case I disable ACK on domoticz side

**Comment by markcame** [comment]
I see that Mysensors send back the echo of the command i will check if the syntax is correct, for now with ack disabled seems good.
The issue not happens with regular non sleeping nodes no error of not received commandfrom node.
it's strange

**Comment by markcame** [comment]
i've look at the code and make a bit of debugging with visual studio what i see is that actually the ack is received but is processed only after the funcion send was performed and so never trigger to true.
i dont't know if it is related to the multithread precedence in the functions if anyone could help i can make some troubleshooting
thanks

**Comment by markcame** [comment]
Finally an explanation of because not work corrrectly the use of ack is ,

when entering in  "case I_PRE_SLEEP_NOTIFICATION:" was performed some action and finally was called the function SendNodeSetCommandImpl that send all potentially buffered messages.

the problem is that this function when is called block the execution of parsing next line until it is finished, so not parsing anymore incoming messages bring to a non ack recived until the function end, but at this point is too late and all timeout is expired.
at this point the ack is parsed and apper on log but is unusable.

@gizmocuz  it is possible to launch the function SendNodeSetCommandImpl in a separate thread to not blocking tha calling function ? i'm not expert of thread managing in c++

**Comment by kofec** [comment]
@markcame I found on forum that it is not allowed to use send in received function in mysensors. And it was my mistake. Maybe it will be helpful for you

**Comment by markcame** [comment]
@kofec sorry but i'm not referring to mysensors send and recive function on the node sketch, but 
rather to the method on how the messages that arriving from the gateway are parsed.
In domoticz the parsing function cannot be blocked calling an external function that send the buffered messages.
With the smartsleep the  notification that a node goes to sleep must  trigger the send of buffered messages, but with the actual implementation this maybe a problem because block the parsing of new messages including the ack

**Comment by waltervl** [comment]
Is this still an issue?

**Comment by MarcusKlerks** [comment]
I think the issue still exists (nothing has changed since then) but I worked around it. I planned to use it on a battery operated device (airco control) where smartsleep is very useful (saves batteries) but since the problem remained I decided to apply adapter power. Don't know what the other people did to solve it.

**Comment by gizmocuz** [comment]
@markcame , yes you could send it with a thread like

				std::thread SendThread([this, NodeID, ChildID, messageType, SubType, Payload, bUseAck, AckTimeout] { SendNodeSetCommandImpl(NodeID, ChildID, messageType, SubType, Payload, bUseAck, AckTimeout); });
				SendThread.detach();

**Comment by waltervl** [comment]
Can someonewith a mysensor with smart sleep create a PR for this?
